### PR TITLE
printer networking rework

### PR DESCRIPTION
**Pull request by marekmosna-prusa3d**
Added features:

- PrusaLink v2.0.0
  - OctoPrint API compatible
  - X-API-Key authentication
  - upload gcodes and start printing from PrusaSlicer
- ESP8266 as additional network device
  - ESP firmware compatible version v1.6
  - ability to re-flash ESP firmware

### CODE REVIEW

**Review by vorner — CHANGES_REQUESTED** [review]
I've went over it, going less deeper in code that seems to have been imported. Some of the notes I have are stylistic or suggestions for minor readability improvements. I don't insist on addressing them, but it seems a good opportunity when they were already identified.

Nevertheless, some of the notes are quite severe ‒ undefined behaviors, possible security issues, etc. I believe these should not go to master before being fixed.

Apart from the in-code comments, I have few more general concerns that don't apply to specific line of code here:

The importe ESP related code seems to be scattered over lib and include, not together. Is that intended?

The imported ESP code seems to be a bit heavy:
* There's a lot of the modules we probably don't use. Yet, we still seem to compile them. I hope the compilation process throws the code out during linking.
* It contains its own memory allocator, its own routine for parsing numbers, … that probably duplicates code we already have.

We have some kind of command-line utility for full-featured computers scattered in the middle of the buddy-board code. Do we need that utility? If so, can it be moved to some other place (maybe the utils directory)?

Similarly, there's some kind of lib/ESP/prusa/snippets. What are they for? Why are they part of `lib`?

The branch adds a lot of `_dbg` messages. Is that OK? Should they stay?

All the scripts in lib/Middlewares/Third_Party/littlefs/scripts got their executable flag removed. Why?

There's a discussion about going with some other approach for using ESP than the AT firmware. The changes now are bit intrusive, they are through a lot of surrounding code and if we decide to go with the other solution, we'll be trying to separate them out later on. I'm not sure what to do about that ‒ unfortunately, the branch contains two major features ‒ ESP and some more API endpoints for the server. I guess somehow filtering through the branch and splitting into two separate branches so they can be reviewed, merged and possibly reverted independently would be quite a lot of work.

The `fsdata_wui_local.c` seems to be a generated file (it certainly looks huge and not written by hand). How is it generated? Why is it in git and not made during the compilation process?

I'm not the one to make the call what to do about these things here, but I think I should share them.

File: lib/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_hal_eth.h
Comment (joshymjose) [reply lib/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_hal_eth.h:678]: why do you change this value?

File: lib/WUI/httpc/httpc.c
Comment (joshymjose) [reply lib/WUI/httpc/httpc.c:40]: the directory */httpc must be removed. It is left over of some test implementation

File: lib/ESP/prusa/src/system/esp_tcp.c
Comment (joshymjose) [inline lib/ESP/prusa/src/system/esp_tcp.c:14]: Interesting copyright notce

File: include/esp/esp_config.h
Comment (DRracer) [reply include/esp/esp_config.h:3]: yes it is

File: include/esp/lwesp_ll_buddy.h
Comment (DRracer) [reply include/esp/lwesp_ll_buddy.h:4]: this is the correct approach, gcc can be quite creative when mixing the two principles incorrectly

File: lib/WUI/alsockets.c
Comment (DRracer) [reply lib/WUI/alsockets.c:109]: generally we try to avoid extern data/structs, so the accessor function is a better solution IMO (makes things not compile if you change it while changing data may compile and even link in some cases just with a warning)

**Review by vorner — DISMISSED** [review]
From MPV, I think this can be merged provided the other things get resolved later on.

**Review by DRracer — APPROVED** [review]
unwanted attribute changes have been reverted as well, approving

**Review by vorner — APPROVED** [review]
OK, but…

Could we _not_ get another „300 files changed“ changeset in this particular branch before it gets merged? This is borderline unreviewable.

**Pull request merged**
